API endpoints

iantance opened the pull request:
Addresses: [ECI-760](https://dripcom.atlassian.net/browse/ECI-760)

I think this is ready for a look. It's branched off of @will-in-wi's  work in progress. Adds an endpoint to set account param and integration token. Adds an endpoint to return status.

will-in-wi commented on Api/Settings.php:
```diff
@@ -0,0 +1,60 @@
+
+class Settings implements SettingsInterface
+{
+		/** @var \Drip\Connect\Model\ConfigurationFactory */
```
This whole file has somewhat goofed up indentation. Could this be fixed? We really need to get a linter on this project.

will-in-wi commented on Model/Configuration.php:
```diff
@@ -4,11 +4,13 @@
     const MODULE_ENABLED_PATH = 'dripconnect_general/module_settings/is_enabled';
     const SALT_PATH = 'dripconnect_general/module_settings/salt';
     const LOG_SETTINGS_PATH = 'dripconnect_general/log_settings';
+		const API_TIMEOUT_PATH = 'dripconnect_general/api_settings/timeout';
```
Do we need a dynamically set timeout? Seems like something we could set as a constant and not as a setting.

iantance replied:
`setTimeout` is still getting called somewhere, so I put this back so that things would load and I could run tests. I agree we don't need it anymore.

will-in-wi commented on Model/Configuration.php:
```diff
@@ -116,4 +152,34 @@ protected function getStoreConfig($path)
+						$path,
+						$val,
+						\Magento\Store\Model\ScopeInterface::SCOPE_WEBSITES,
+						$this->storeId
```
This won't work… Store ID and website ID aren't the same, so this'll return invalid data.

Might be worth fixing in a different PR, however, where we switch over to just having everything be based on the website scope instead of store view.

will-in-wi commented on etc/adminhtml/system.xml:
```diff
@@ -27,9 +27,12 @@
             <group id="api_settings" type="text" sortOrder="10" showInDefault="1" showInWebsite="1" showInStore="1" translate="label">
                 <label>API Settings</label>
-                <field id="integration_token" translate="label" type="text" sortOrder="21" showInDefault="1" showInWebsite="0" showInStore="0">
+                <field id="integration_token" translate="label" type="text" sortOrder="21" showInDefault="1" showInWebsite="1" showInStore="1">
```
I don't think we want to set this at the store level at all…
```suggestion
                <field id="integration_token" translate="label" type="text" sortOrder="21" showInDefault="1" showInWebsite="1" showInStore="0">
```

will-in-wi commented on etc/adminhtml/system.xml:
```diff
@@ -27,9 +27,12 @@
+                <field id="integration_token" translate="label" type="text" sortOrder="21" showInDefault="1" showInWebsite="1" showInStore="1">
                     <label>Drip Integration Token</label>
                 </field>
+                <field id="account_param" translate="label" type="text" sortOrder="25" showInDefault="1" showInWebsite="1" showInStore="1">
```
I don't think we want this set at the store level.
```suggestion
                <field id="account_param" translate="label" type="text" sortOrder="25" showInDefault="1" showInWebsite="1" showInStore="0">
```

will-in-wi commented on devtools_m2/cypress/integration/RestApi.feature:
```diff
@@ -0,0 +1,24 @@
+
+  Scenario: Authorized request to set integration token
+    Given I am logged into the admin interface
+      And I have configured Drip to be enabled for 'main'
```
I think most of these tests might be operating at a store level and giving us bad results. We might want to do a refactor PR to switch everything to work at a website level, although that could happen after this one (or before, with a rebase…)

iantance commented:
@will-in-wi This one is ready for another look.

I wasn't sure how we might want to handle the `is_enabled` setting on the plugin. I set this up to enable it when the token and account_param are set, but let me know if you think it would be better to require that people enable it in the Magento UI.

will-in-wi commented on Api/Settings.php:
```diff
@@ -0,0 +1,55 @@
+    /**
+     * {@inheritdoc}
+     */
+    public function updateSettings($websiteId, $accountParam, $integrationToken) {
```
Might want to explicitly handle the case where the website id doesn't exist. And possibly also the case where we're setting these settings on the default website instead of the specific one.

will-in-wi commented on Api/SettingsInterface.php:
```diff
@@ -0,0 +1,19 @@
+     public function updateSettings($websiteId, $accountParam, $integrationToken);
+
+    /**
+     * POST for integration settings API
```
Should this be a GET? It has no side-effects.

will-in-wi commented on Model/Configuration.php:
```diff
@@ -70,6 +71,11 @@ public function getIntegrationToken()
         return $this->getConfig(self::INTEGRATION_TOKEN);
     }
 
+        public function getAccountParam()
```
Some indentation weirdness here.

will-in-wi commented on Model/Configuration.php:
```diff
@@ -78,6 +84,11 @@ public function isEnabled()
 
+	public function enable()
+	{
+		return $this->setConfig(self::MODULE_ENABLED_PATH, 1);
```
Indentation weirdness here.

Edit: I'm going to stop flagging these. It looks like we have a mix of tabs and spaces all over this diff.

iantance commented:
@will-in-wi Ready for another look.

I removed that `enabled` setting.
Updates and tests to handle default scope and missing website.
Cleaned up indentation (I think).
